### PR TITLE
Geometry_Engine: #924 Add Interfaces to Boolean Intersection / Union / Difference

### LINKED ISSUE #924: Geometry_Engine: Add interface methods for the Boolean operations on regions
As discussed in #920, interface methods are needed for the Boolean operations to allow working on `Polylines` and `PolyCurves` at the same time.

An option could be making the Boolean operations work on `ICurves` as single generic methods.

### PR DESCRIPTION
**Pull request by LMarkowski**
### Issues addressed by this PR

closes #924 

Added Interfaces for boolean methods. Now they work on both PolyCurves and Polylines as one input. Circles are are also implemented as singular Circles and as a parts of a PolyCurves. To be discussed is if we need separate Polyline, PolyCurve and Interface methods. I think we could have one method for Polylines as it is faster than others and one universal method for the rest (including Polylines). Method for only PolyCurves is a bit redundant.
I also did some updates in all boolean methods as there appeared new egde cases :wink:

However there are some issues with performance. In some cases BhoM methods are much faster then GH's, but they also can be slower. I give some examples in the thest file.

### Test files
Regular [Intersection](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Geometry_Engine/Compute/BooleanIntersectionV3.gh?csf=1&e=BY2efi)
Regular [Union](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Geometry_Engine/Compute/BooleanUnionV3.gh?csf=1&e=24Lmd4)
Regular [Difference](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Geometry_Engine/Compute/BooleanDifferenceV3.gh?csf=1&e=9jxJpK)
[Performence issues](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Geometry_Engine/Geometry_Engine-issue924-BooleanInterfacesPerformanceIssues/Geometry_Engine-issue924-BooleanInterfacesPerformanceIssues.gh?csf=1&e=SDehQZ)
Regular [SplitAtPoints](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Geometry_Engine/Modify/SplitAtPoints.gh?csf=1&e=65S2CW)

#### Added:
- `Compute.IBooleanIntersection()` method added in the `Geometry_Engine` for `ICurve` and `ICurve` classes
- `Compute.IBooleanIntersection()` method added in the `Geometry_Engine` for `List<ICurve>` class
- `Compute.IBooleanUnion()` method added in the `Geometry_Engine` for `List<ICurve>` class
- `Compute.IBooleanDifference()` method added in the `Geometry_Engine` for `ICurve` and `List<ICurve>` classes

#### Changed:
- `Compute.BooleanIntersection()` method changed in the `Geometry_Engine` for `Polyline` and `Polyline` classes
- `Compute.BooleanUnion()` method changed in the `Geometry_Engine` for `List<Polyline>` class
- `Compute.BooleanDifference()` method changed in the `Geometry_Engine` for `Polyline` and `List<Polyline>` classes
- `Compute.BooleanIntersection()` deprecated method in the `Geometry_Engine` for `PolyCurve` and `PolyCurve` classes
- `Compute.BooleanUnion()` deprecated method in the `Geometry_Engine` for `List<PolyCurve>` class
- `Compute.BooleanDifference()` deprecated method in the `Geometry_Engine` for `PolyCurve` and `List<PolyCurve>` classes

### Additional comments
As said above, I think maybe methods for PolyCurves can be erased. The Interfaces would be our universal methods. Maybe without `I` as they are not interfaces by definition.

### CODE REVIEW

**Comment by pawelbaran** [comment]
Plus could you please change the name of the PR? No info on which project does it influence.

File: Geometry_Engine/Compute/BooleanIntersection.cs [hunk 1]
```diff
@@ -298,5 +350,150 @@ public static List<PolyCurve> BooleanIntersection(this List<PolyCurve> regions,
...
+            }
+
+            return result;
+        }
+
+        /***************************************************/
+        /***          Private methods                    ***/
+        /***************************************************/
+        
+        private static Boolean IsGeometricallyEqual(this ICurve curve, ICurve refCurve, double tolerance = Tolerance.Distance)
```
[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply Geometry_Engine/Compute/BooleanIntersection.cs]: https://github.com/BHoM/BHoM_Engine/blob/master/Environment_Engine/Query/IsIdentical.cs

Take a look at that method, that's how we decided to check for identical geometry on Environment `Panel` - might be worth stealing into here and then I can modify that method to just call this one?

**Comment by pawelbaran** [comment]
Still unhappy with the name: my concern was that it does not start with _Geometry_Engine: (...)_

**Comment by LMarkowski** [comment]
> Still unhappy with the name: my concern was that it does not start with _Geometry_Engine: (...)_

🤦‍♂  ...sorry

File: Geometry_Engine/Compute/BooleanIntersection.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (LMarkowski) [reply Geometry_Engine/Compute/BooleanIntersection.cs]: I get your point. I need this method to cull (or sometimes not) overlapping edges. At first I was using IsEqual but it was returning nulls for elements of different types what was not acceptable. The possible inputs of my method are segments of regions' edges, split by their intersection points. There is no way in which they could intersect so it is enough to check only 3 points. It as well works for segments being PolyCurve and PolyLines.
I can change it into your way, but now I think it is as light as possible and the methods are already heavy.

[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply Geometry_Engine/Compute/BooleanIntersection.cs]: In that case I would argue the method might not be appropriately named - the way I interpreted it was a check that 2 curves are equal, and those 2 curves can be any length, any shape, etc., so a more definitive check of each point should be made.
Of course there is also a philosophical question about equality - if we have two straight lines for example,  both start and end in the same spot, but one of the lines has an extra middle point geometrically (for whatever reason), can they be considered equal? Of course with the method @pawelbaran and I propose, they would be considered not-equal, because one line has 3 points to the other line which has 2.
Of course that's where a combination of methods can then be used, like the `CleanPolyline` method I put in a few weeks back which would remove that extra point and allow the two lines to be geometrically equal on the discussed methods.

Anyway, I digress...

TL;DR - rename the method if the method is not actually checking geometrical equality - I think that's the summation of that :wink: @pawelbaran would you agree?

[the remark this answers is not in the document]
Comment (LMarkowski) [reply Geometry_Engine/Compute/BooleanIntersection.cs]: Ok ok ok ok.. What I meant is that this method checks geometrical equality but **only** in case of my specific method. Because of how the input is created it's elements can't intersect, so if startpoint, midpoint and endpoint are equal that means that elements are equal. 😄  I should have hidden it somewhere in the code so there would be no problem 😄 
Renaming in progres 😉

[the remark this answers is not in the document]
Comment (pawelbaran) [reply Geometry_Engine/Compute/BooleanIntersection.cs]: Yeah, agreed with @FraserGreenroyd. My suggestion would be to name it with `IsSimilarSegment` etc.

**Comment by LMarkowski** [comment]
Summarizing:
- 80 errors for `BooleanIntersection(PolyCurves)` were caused by a little bug in `SplitAtPoints` which have been fixed. (I added test file for `Split`).
- 2 errors (actually one - test for PolyCurves was duplicated) for `BooleanDiffernce(PolyCurves)` are caused by difference between results of GH and BHoM, which is slightly greater than tolerance, but still around 1e-6.
- `IsGeometricallyEqual` has been lightened and renamed into `IsSimilarSegment`
- `BooleanIntersection(Circle, PolyCurve)` has been removed as it is useless and was a leftover from the previous concept.

**Comment by LMarkowski** [comment]
Update:
- Totally agree that `I` is inadequate. Removed.
- Change input type from `List<ICurve>` to `IEnumberable<ICurve>`. Done.
- We already have a talk about it with @pawelbaran and had same conclusions. As we all are agreed I deprecated the PolyCurves methods. Now they call ICurve methods.

I also changed the returning type from ICurve to PolyCurve. We decided that it's more reasonable.
I've updated the test files to new methods.

**Comment by IsakNaslundBh** [comment]
LGTM now, and as Pawel has approved as well, I assume any impact on other toolkits have been checked. Happy to merge 👍
